core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -408,6 +413,38 @@ private Expression unwrapCast(ComparisonExpression expression)
             return new ComparisonExpression(operator, cast.getExpression(), literalEncoder.toExpression(literalInSourceType, sourceType));
         }
 
+        private Optional<Expression> unwrapTimestampToDateCast(TimestampType sourceType, DateType targetType, ComparisonExpression.Operator operator, Expression timestampExpression, long dateLiteral)
```

Comment:
Writing a dedicated case for every type is less than ideal. Ideally, this kind of rewrite should work for any pair of types (also plugged by connectors). Hence, IMO a better approach would be to introduce `saturated_ceil_cast` in addition to `saturated_floor_cast` and solve this problem is a more generic way. This class already seem to have an awful lot of tricky edge cases (e.g timestamps, decimals, doubles)

Reply:
Yeah - having a generic way is an ultimate goal (I should have stated that in the commit description). Not 100% sure that saturated_ceil_cast is the best (only) approach, but seems like one of the possible solutions. Does it mean we do not have tactical improvement for the period, when we do not have a generic solution? I do not feel it adds much complexity to be honest.

Reply:
The other thing is that (it looks so) that we currently do not define even saturated_floor cast `date -> timestamp` (neither most other type pairs), so just reworking this to saturated casts will not solve the problem. Scope is wider.

Reply:
> Does it mean we do not want to have tactical improvement for the period

I'm fine with tactical improvement. However, saturated cast approach should not be that much bigger IMO. We would need:
1. Saturated casts for `date -> timestamp`
2. New unwrap rule that uses them (or maybe extending this one). Seems like a little bit more work that introduced more generic mechanism.